### PR TITLE
mds: multimds snap compatibility

### PR DESCRIPTION
**Pull request by ukernel**
1. disable multimds by default for filesystem created by pre-mimic mds.
2. mds automatically enable multimds if it detects there is no old format snapshot. (either never created or all are deleted)
3. mds automatically enable multimds after whole filesystem scrub finishes

### CODE REVIEW

**Comment by batrick** [comment]
http://pulpito.ceph.com/?branch=wip-pdonnell-testing-20180425.144904

**Comment by batrick** [comment]
@ukernel http://pulpito.ceph.com/pdonnell-2018-04-25_21:10:04-multimds-wip-pdonnell-testing-20180425.144904-testing-basic-smithi/2439053/

File: PendingReleaseNotes [hunk 1]
```diff
@@ -63,6 +63,23 @@
...
+    To enable snapshot on existing filesystems, use command:
+      - ceph fs set <fs_name> allow_new_snaps
+
+    The on-disk format of snapshot metadata has changed. The old format
+    metadata can not be properly handled in multiple active MDS configuration.
+    To guarantee all snapshot metadata on existing filesystems get updated,
+    perform the following upgrade sequence:
+
+    Reduce the number of ranks to 1 (`ceph fs set <fs_name> max_mds 1`),
+    deactivate all other ranks (`ceph mds deactivate <fs_name>:<n>`),
```
Comment (batrick) [inline PendingReleaseNotes]: Best to refer the admin to the upgrade procedure document for the MDS parts: http://docs.ceph.com/docs/luminous/cephfs/upgrading/#upgrading-the-mds-cluster

Just add a note to skip the step which resets max_mds to its previous setting. Then they can proceed with the scrubbing commands and finally resetting max_mds to the desired setting.

File: PendingReleaseNotes [hunk 2]
```diff
@@ -63,6 +63,23 @@
...
+    upgrade/restart MON/OSD, then upgrade/restart MDS. The multiple active
+    MDS feature is now disabled by the MON. To re-enable the feature, either
+    delete all old snapshots or scrub the whole filesystem:
+      - ceph daemon <mds of rank 0> scrub_path / force recursive repair
+      - ceph daemon <mds of rank 0> scrub_path ~mdsdir force recursive repair
```
Comment (batrick) [inline PendingReleaseNotes]: `~mdsdir` needs shell escaped or surrounded with single quotes.

**Comment by ukernel** [comment]
The last commit is fix for the failure

**Comment by ukernel** [comment]
The last commit is fix for the failure

**Comment by batrick** [comment]
http://pulpito.ceph.com/?branch=wip-pdonnell-testing-20180427.172333

**Comment by batrick** [comment]
Zheng, I tried to do some ad-hoc tests with this branch and didnt' get what I expected:

First I built `ebd0fcd0674e6bc527a7af548bd420db943f98a6^`  (commit prior to your snapshot fixes) and started a vstart.sh cluster, ceph-fuse client, and made some snapshots:

````
pdonnell@senta02 ~/ceph/build$ D=$(mktemp -d -p ~/mnt/); mkdir -p "$D"; run sudo bin/ceph-fuse -f --id fs -c ceph.conf "$D"
sudo bin/ceph-fuse -f --id fs -c ceph.conf /home/pdonnell/mnt/tmp.owGSed6b5b
2018-04-27 00:19:00.551 7f17c5ac9c00 -1 WARNING: all dangerous and experimental features are enabled.
2018-04-27 00:19:00.559 7f17c5ac9c00 -1 WARNING: all dangerous and experimental features are enabled.
2018-04-27 00:19:00.563 7f17c5ac9c00 -1 WARNING: all dangerous and experimental features are enabled.
2018-04-27 00:19:00.563 7f17c5ac9c00 -1 init, newargv = 0x89ac12afc0 newargc=7
ceph-fuse[12790]: starting ceph client
ceph-fuse[12790]: starting fuse
````

````
pdonnell@senta02 ~/ceph$ cd /home/pdonnell/mnt/tmp.owGSed6b5b
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b$ ls
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b$ ls
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b$ mkdir foo
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b$ cd foo
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo$ ls -lha
total 1.0K
drwxr-x--- 1 pdonnell pdonnell 0 Apr 27 17:29 .
drwxr-xr-x 1 pdonnell pdonnell 0 Apr 27 17:29 ..
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo$ mkdir .snap/test
mkdir: cannot create directory ‘.snap/test’: Operation not permitted
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo$ mkdir .snap/test
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo$ ls -lha ^C
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo$ touc^C
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo$ echo hi > bar
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo$ mkdir .snap/test1
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo$ ls -lha
total 1.5K
drwxr-x--- 1 pdonnell pdonnell 3 Apr 27 17:29 .
drwxr-xr-x 1 pdonnell pdonnell 0 Apr 27 17:29 ..
-rw-r----- 1 pdonnell pdonnell 3 Apr 27 17:29 bar
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo$ mkdir foo
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo$ cd foo
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo/foo$ ls
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo/foo$ ln ../bar
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo/foo$ rm bar
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo/foo$ ls
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo/foo$ ln ^C
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo/foo$ echo hello > lkasjdf
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo/foo$ mkdir .snap/^C
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo/foo$ mv ../bar  .
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo/foo$ cat ../.snap/test
test/  test1/
pdonnell@senta02 ~/mnt/tmp.owGSed6b5b/foo/foo$ cat ../.snap/test1/bar
hi
````

Then I checked out and built this branch, `72d249ab0a2bbdef3a7c4259a61b299a8961cbc4`, performed the upgrade procedure of restarting mons, osds, then mdss. I never had 2 active MDS. The ceph-fuse client was not upgraded.

After completing the upgrade, I tried to increase max_mds:

````
$ bin/ceph fs set cephfs_a max_mds 2
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2018-04-27 17:37:50.688 7eff7cd73700 -1 WARNING: all dangerous and experimental features are enabled.
2018-04-27 17:37:50.724 7eff7cd73700 -1 WARNING: all dangerous and experimental features are enabled.
$ bin/ceph status
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2018-04-27 18:02:26.555 7fb1e0456700 -1 WARNING: all dangerous and experimental features are enabled.
2018-04-27 18:02:26.575 7fb1e0456700 -1 WARNING: all dangerous and experimental features are enabled.
  cluster:
    id:     ffc40fc3-33ba-4c30-b83b-1ff3fcc01101
    health: HEALTH_WARN
            2 slow ops, oldest one blocked for 1278 sec
 
  services:
    mon: 3 daemons, quorum a,b,c
    mgr: x(active)
    mds: cephfs_a-2/2/2 up  {0=a=up:active,1=c=up:active}, 1 up:standby
    osd: 3 osds: 3 up, 3 in
 
  data:
    pools:   2 pools, 16 pgs
    objects: 47 objects, 212KiB
    usage:   3.19GiB used, 27.0GiB / 30.2GiB avail
    pgs:     16 active+clean
$ bin/ceph fs dump --format=json
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2018-04-27 18:02:12.856 7ff6b12fe700 -1 WARNING: all dangerous and experimental features are enabled.
2018-04-27 18:02:12.880 7ff6b12fe700 -1 WARNING: all dangerous and experimental features are enabled.
dumped fsmap epoch 21

{"epoch":21,"default_fscid":1,"compat":{"compat":{},"ro_compat":{},"incompat":{"feature_1":"base v0.20","feature_2":"client writeable ranges","feature_3":"default file layouts on dirs","feature_4":"dir inode in separate object","feature_5":"mds uses versioned encoding","feature_6":"dirfrag is stored in omap","feature_8":"no anchor table","feature_9":"file layout v2","feature_10":"snaprealm v2"}},"feature_flags":{"enable_multiple":false,"ever_enabled_multiple":false},"standbys":[{"gid":34152,"name":"b","rank":-1,"incarnation":0,"state":"up:standby","state_seq":2,"addr":"127.0.0.1:6868/2430442436","standby_for_rank":-1,"standby_for_fscid":-1,"standby_for_name":"","standby_replay":false,"export_targets":[],"features":2305244844817448955,"epoch":15}],"filesystems":[{"mdsmap":{"epoch":21,"flags":30,"ever_allowed_features":18,"explicitly_allowed_features":18,"created":"2018-04-27 00:18:40.188378","modified":"2018-04-27 17:43:05.766519","tableserver":0,"root":0,"session_timeout":60,"session_autoclose":300,"max_file_size":1099511627776,"last_failure":0,"last_failure_osd_epoch":30,"compat":{"compat":{},"ro_compat":{},"incompat":{"feature_1":"base v0.20","feature_2":"client writeable ranges","feature_3":"default file layouts on dirs","feature_4":"dir inode in separate object","feature_5":"mds uses versioned encoding","feature_6":"dirfrag is stored in omap","feature_8":"no anchor table","feature_9":"file layout v2","feature_10":"snaprealm v2"}},"max_mds":2,"in":[0,1],"up":{"mds_0":44156,"mds_1":34158},"failed":[],"damaged":[],"stopped":[],"info":{"gid_34158":{"gid":34158,"name":"c","rank":1,"incarnation":17,"state":"up:active","state_seq":10,"addr":"127.0.0.1:6869/152276863","standby_for_rank":-1,"standby_for_fscid":-1,"standby_for_name":"","standby_replay":false,"export_targets":[],"features":2305244844817448955},"gid_44156":{"gid":44156,"name":"a","rank":0,"incarnation":10,"state":"up:active","state_seq":5,"addr":"127.0.0.1:6867/1680532985","standby_for_rank":-1,"standby_for_fscid":-1,"standby_for_name":"","standby_replay":false,"export_targets":[],"features":2305244844817448955}},"data_pools":[1],"metadata_pool":2,"enabled":true,"fs_name":"cephfs_a","balancer":"","standby_count_wanted":1},"id":1}]}
$ for s in a b c; do bin/ceph daemon mon.$s version; done
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
{"version":"13.0.2-1535-g72d249a","release":"mimic","release_type":"dev"}
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
{"version":"13.0.2-1535-g72d249a","release":"mimic","release_type":"dev"}
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
{"version":"13.0.2-1535-g72d249a","release":"mimic","release_type":"dev"}
$ for s in a b c; do bin/ceph daemon mds.$s version; done
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
{"version":"13.0.2-1535-g72d249a","release":"mimic","release_type":"dev"}
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
{"version":"13.0.2-1535-g72d249a","release":"mimic","release_type":"dev"}
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
{"version":"13.0.2-1535-g72d249a","release":"mimic","release_type":"dev"}
````

I think this should have failed when setting `max_mds` to 2, right?

I will try to whip together a quick QA test that verifies this by installing luminous then does an upgrade.

**Comment by ukernel** [comment]
please 'ceph fs dump', I need check filesystem's flags

**Comment by ukernel** [comment]
failures of 'sudo rm -rf -- /home/ubuntu/cephtest/mnt.0/client.0/tmp'  are likely caused by kernel client bug

**Comment by batrick** [comment]
http://pulpito.ceph.com/pdonnell-2018-05-03_01:23:04-fs:upgrade-wip-pdonnell-pr21374-distro-basic-mira/

**Comment by batrick** [comment]
http://pulpito.ceph.com/pdonnell-2018-05-03_04:03:45-fs:upgrade-wip-pdonnell-pr21374-distro-basic-mira/

Another run, fixed the client upgrade. (forgot to restart the client)

**Comment by batrick** [comment]
Hrm, can't just restart the client like a daemon. I'll fix this in the morning.

**Comment by batrick** [comment]
http://pulpito.ceph.com/pdonnell-2018-05-03_16:09:55-fs:upgrade-wip-pdonnell-pr21374-distro-basic-mira/

File: src/mon/FSCommands.cc [hunk 3]
```diff
@@ -361,12 +361,6 @@ class SetHandler : public FileSystemCommandHandler
         });
 	ss << "disabled new snapshots";
       } else {
-	string confirm;
-	if (!cmd_getval(g_ceph_context, cmdmap, "confirm", confirm) ||
-	    confirm != "--yes-i-really-mean-it") {
```
Comment (tchaikov) [inline src/mon/FSCommands.cc]: this breaks https://github.com/ceph/ceph/blob/master/qa/workunits/cephtool/test.sh#L991

**Pull request merged**